pkg/serializer/internal/metrics/events.go

```diff
@@ -280,3 +292,240 @@ func (events Events) CreateMarshalersBySourceType() []marshaler.StreamJSONMarsha
 	}
 	return values
 }
+
+var jsonConfig = jsoniter.Config{}.Froze()
+
+// eventsMarshaler2 serializes events to json grouped by source type name. namely:
+//
+//	{ "events": { "source type": [ { event }, ... ], ... }, ... }
+type eventsMarshaler2 struct {
+	logger      log.Component
+	compression compression.Component
+
+	events []*event.Event
+	stream *jsoniter.Stream
+
+	bufferContext  *marshaler.BufferContext
+	compressor     *stream.Compressor
+	header, footer []byte
+
+	// last source type name successfully added to the compressor
+	sourceTypeName string
+	// if any events were added to the compressor since starting new payload
+	eventsInPayload bool
+
+	maxPayloadSize      int
+	maxUncompressedSize int
+
+	payloads []*transaction.BytesPayload
+}
+
+// MarshalEvents serializes an array of events into one or more compressed intake payloads.
+func MarshalEvents(
+	events event.Events,
+	hostname string,
+	config config.Component,
+	logger log.Component,
+	compression compression.Component,
+) ([]*transaction.BytesPayload, error) {
+	m := createMarshaler2(events, hostname, config, logger, compression)
+	return m.marshal()
+}
+
+func createMarshaler2(
+	events event.Events,
+	hostname string,
+	config config.Component,
+	logger log.Component,
+	compression compression.Component,
+) *eventsMarshaler2 {
+	events = slices.Clone(events)
+	sort.Slice(events, func(i, j int) bool {
+		return events[i].SourceTypeName < events[j].SourceTypeName
+	})
+
+	stream := jsoniter.NewStream(jsonConfig, nil, 4096)
+
+	writeEventsHeader(stream)
+	header := slices.Clone(stream.Buffer())
+	stream.Reset(nil)
+
+	stream.WriteArrayEnd()
+	_ = writeEventsFooter(stream, hostname)
+	footer := slices.Clone(stream.Buffer())
+	stream.Reset(nil)
+
+	return &eventsMarshaler2{
+		events:      events,
+		logger:      logger,
+		compression: compression,
+
+		stream: stream,
+
+		bufferContext: marshaler.NewBufferContext(),
+		header:        header,
+		footer:        footer,
+
+		maxPayloadSize:      config.GetInt("serializer_max_payload_size"),
+		maxUncompressedSize: config.GetInt("serializer_max_uncompressed_payload_size"),
+	}
+}
+
+func (e *eventsMarshaler2) flushPayload() error {
+	err := e.closePayload()
+	if err != nil {
+		return nil
+	}
```

Comment:
What would be the reason of this failing? Should we log something?
Also, returning `nil` looks suspicious, is it a mistake?